### PR TITLE
Support an explicit annotation for marker traits

### PR DESCRIPTION
**Pull request by scottmcm**
From the tracking issue for rust-lang/rfcs#1268:
> It seems obvious that we should make a `#[marker]` annotation. ~ https://github.com/rust-lang/rust/issues/29864#issuecomment-368959441

This PR allows you to put `#[marker]` on a trait, at which point:
- [x] The trait must not have any items ~~All of the trait's items must have defaults~~
- [x] Any impl of the trait must be empty (not override any items)
- [x] But impls of the trait are allowed to overlap

r? @nikomatsakis

### CODE REVIEW

**Comment by scottmcm** [comment]
I think this is functional and sound, but it's missing the "this trait would be impossible to implement" error when defining a trait without default implementations for its items.

I'd appreciate advice on where that should go, as well as any other feedback on how I've been going about this.

**Comment by nikomatsakis** [comment]
@scottmcm this is confusing to me. I thought a marker trait had no items -- not "items with defaults".

**Comment by scottmcm** [comment]
@nikomatsakis See https://github.com/rust-lang/rust/issues/29864#issuecomment-368478686 for some context.  But yes, it's an extension beyond the RFC, so I'm fine not implementing that for now.

I do still need advice on where to put the "has no items" valuation check, though.  (Once I know where I can figure out how to get the item list.)

**Comment by nikomatsakis** [comment]
@scottmcm I see. That seems like an ok extension, actually, but in that case I don't know what this is referring to...

> I think this is functional and sound, but it's missing the "this trait would be impossible to implement" error when defining a trait without default implementations for its items.

...oh, I see, I thought you meant some fancy check designed to prove that the trait is imposible to implement. But actually you just mean that you want to report an error if you label a trait as `#[marker]` and it has items that lack defaults?

In that case, I think we should do that .. hmm .. either in WF or in coherence.  Coherence seems like -- today -- it is focused on checking properties of *impls* so maybe wfcheck would be a better place. 

Basically extending this function:

https://github.com/rust-lang/rust/blob/70a21e89f136c91f676959c126666d13a2deeadf/src/librustc_typeck/check/wfcheck.rs#L305-L311

We should probably do a quick lang-team FCP here too.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

This PR modifies the special treatment of marker traits in two ways. First, you have to declare the traits explicitly, by writing `#[marker]`, but secondly, it permits items in those traits -- so long as those items have defaults. However, once a trait is defined as `#[marker]`, impls of the trait are then disallowed from containing any items, so you know that the defaults from the trait will be used.

Seems reasonable to me.

**Comment by withoutboats** [comment]
@rfcbot concern vague

I'm :+1: on having a `#[marker]` annotation for marker traits with no items, and requiring that to get the overlapping allowance as a statement of your indefinite, permanent intent that this trait is a marker trait.

I'm not so convinced that redefining marker trait to include traits with default items and then adding additional checks that those are never overriden is a good idea. It comes at a nontrivial cost of complexity for users in terms of understanding the meaning and mechanics of *both* default methods and marker traits. The motivation doesn't seem that strong to me.

**Comment by Centril** [comment]
I very much agree that the incoherency should be opt-in rather than be implicit.

However; this was not discussed in the RFC (unless I somehow missed that but I did look at every comment...) and this could use some more elaborate and detailed justification.
For example, the changes wrt. provided definitions could use some more explaining per @withoutboats's concern. This is also, relatively speaking to the size of the original RFC, quite a large change.
Therefore I'd like to move this back to the drawing board, so:

@rfcbot concern needs-rfc

In addition; I am not so sure about the concrete syntax here.
For example, it seems to me that `#[overlappable]` (as suggested by @Vurich in https://github.com/rust-lang/rust/issues/29864#issuecomment-368480582) or `#[incoherent]` is more informative than `#[marker]` as to what effect it has. I think this holds particularly if associated items with defaults are permitted but can't be overridden as it isn't a marker trait anymore.

The choice of using an attribute rather than some keyword as a prefix to the trait (e.g. `overlappable trait Foo { .. })` hasn't been sufficiently justified either in my opinion. The justification that has been given (https://github.com/rust-lang/rust/issues/29864#issuecomment-357081249) is merely that there is precedent to use an attribute. But precedent alone is not sufficient. In would like to see a discussion of this in an RFC. As such:

@rfcbot concern concrete-syntax

**Comment by Centril** [comment]
To elaborate a bit: I'm fine with landing this PR and using `#[marker]` as a temporary syntax; but not with resolving the final design at this stage.

**Comment by nrc** [comment]
I agree with @Centril and @withoutboats - this seems like a big change and marker traits with default methods adds considerably to the complexity of the feature, as well as making `marker` not really an appropriate name, IMO.

**Comment by scottmcm** [comment]
Ok, I'll just do the "opt into the RFC behaviour using an attribute" part of the change, making having _anything_ in a `marker` trait illegal and postponing any extensions for later.

I think `#[marker]` is correct because of the [`core::marker`](https://doc.rust-lang.org/std/marker/index.html) module, so this is already a term of art.  This PR isn't stabilizing anything, so keep bikeshedding and we can change it later.

**Comment by withoutboats** [comment]
@rfcbot resolve vague

**Comment by Centril** [comment]
@rfcbot resolve needs-rfc
@rfcbot resolve concrete-syntax

Let's get this PR through with `#[marker]` as temporary syntax :)
But before stabilization I'd like to see an RFC.
I'd also prefer a different naming; I'll elaborate on why in the tracking issue.

File: src/librustc/hir/check_attr.rs [hunk 1]
```diff
@@ -114,6 +118,31 @@ impl<'a, 'tcx> CheckAttrVisitor<'a, 'tcx> {
...
+                    .emit();
+                return;
+            }
+        }
+
+        if attr.meta_item_list().is_some() || attr.value_str().is_some() {
+            struct_span_err!(self.tcx.sess,
+                             attr.span,
+                             E0714,
+                             "attribute should be empty")
```
Comment (eddyb) [inline src/librustc/hir/check_attr.rs]: Hmm maybe we shouldn't add an error code for every single of these errors...

File: src/librustc/hir/check_attr.rs [hunk 2]
```diff
@@ -114,6 +118,31 @@ impl<'a, 'tcx> CheckAttrVisitor<'a, 'tcx> {
...
+                                 attr.span,
+                                 E0713,
+                                 "attribute can only be applied to a trait")
+                    .span_label(item.span, "not a trait")
+                    .emit();
+                return;
+            }
+        }
+
+        if attr.meta_item_list().is_some() || attr.value_str().is_some() {
```
Comment (eddyb) [inline src/librustc/hir/check_attr.rs]: Is this enough? I know this is most likely copied from somewhere, but it seems the wrong way around (banning a few known attribute forms, instead of checking that it's a specific attribute form). cc @petrochenkov

**Review by eddyb — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@scottmcm ping me when you've updated with the more minimal scope; I agree with the general consensus that we ought to go through RFC process here.

File: src/librustc/hir/check_attr.rs (hunk 2 above)
Comment (petrochenkov) [reply src/librustc/hir/check_attr.rs]: Yeah, `is_word()` should be more appropriate.

Comment (scottmcm) [reply src/librustc/hir/check_attr.rs]: Switched to `is_word`; thanks!

File: src/librustc/hir/check_attr.rs (hunk 1 above)
Comment (scottmcm) [reply src/librustc/hir/check_attr.rs]: Agreed, these ones are a bit too boring to seem worth bothering.  Removed.

**Comment by scottmcm** [comment]
Ok, @nikomatsakis, I've restricted this to the minimal scope:

https://github.com/rust-lang/rust/blob/c991f1134da48c5990239a49d81950b33ee0a3ab/src/test/ui/marker_trait_attr/marker-trait-with-associated-items.rs#L32-L48

**Comment by nikomatsakis** [comment]
@bors r=me with a writeup for relnotes.

@scottmcm can you add a brief paragraph to leave for release notes?

I do not anticipate much comment here, and we can always back this change out (it's not affecting anything stable).

**Comment by scottmcm** [comment]
@nikomatsakis Ok, how about something like this:

> If you're using [`overlapping_marker_traits`](https://doc.rust-lang.org/nightly/unstable-book/language-features/overlapping-marker-traits.html), please migrate your code to [`marker_trait_attr`](https://doc.rust-lang.org/nightly/unstable-book/language-features/marker-trait-attr.html).  The former has the problem that it makes adding a method to an empty trait a potentially-breaking change, so the latter requires that the trait opt-in to supporting overlapping implementations.

But both features continue to work and are unstable, so it's surprising to me that we'd relnote them at all.

**Comment by nikomatsakis** [comment]
@bors r+

I'm not sure if we usually relnote unstable features. Seems like a nice thing to do.
